### PR TITLE
Mark `preact` as externals for server bundle

### PR DESCRIPTION
**Pull request by jackyef**
3rd party packages usually imports `react` or `preact` during the runtime. The server bundle shouldn't keep its own copy of `preact` to avoid issues with more than one instance of `preact` existing in the same runtime.

Example issue caused by duplicate `preact`: https://github.com/preactjs/preact/issues/1959#issuecomment-541087225

### CODE REVIEW

File: examples/using-preact/next.config.js [hunk 1]
```diff
@@ -14,6 +14,11 @@ module.exports = withPrefresh({
       }
     }
 
+    if (isServer) {
+      // mark `preact` stuffs as external for server bundle to prevent duplicate copies of preact
+      config.externals.push(/^preact/);
```
Comment (developit) [inline examples/using-preact/next.config.js:19]: Probably best to use the regex from the chunk configuration above - that was this doesn't end up marking something like `preact-markup` as external:

```suggestion
      config.externals.push(/^(preact|preact-render-to-string|preact-context-provider)([\\/]|$)/);
```

**Review by developit — DISMISSED** [review]
LGTM aside from the regex nit.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| buildDuration | 12s | 12s | ⚠️ +27ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 1.971 | 1.999 | ⚠️ +0.03 |
| / avg req/sec | 1268.25 | 1250.62 | ⚠️ -17.63 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.268 | 1.158 | -0.11 |
| /error-in-render avg req/sec | 1972.05 | 2159.17 | +187.12 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.79 kB | 5.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| index.html gzip | 955 B | 955 B | ✓ |
| link.html gzip | 962 B | 962 B | ✓ |
| withRouter.html gzip | 947 B | 947 B | ✓ |
| Overall change | 2.86 kB | 2.86 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| buildDuration | 13.4s | 13.3s | -99ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.79 kB | 5.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _error.js | 881 kB | 881 kB | ✓ |
| 404.html | 4.17 kB | 4.17 kB | ✓ |
| hooks.html | 3.79 kB | 3.79 kB | ✓ |
| index.js | 881 kB | 881 kB | ✓ |
| link.js | 923 kB | 923 kB | ✓ |
| routerDirect.js | 917 kB | 917 kB | ✓ |
| withRouter.js | 917 kB | 917 kB | ✓ |
| Overall change | 4.53 MB | 4.53 MB | ✓ |

</details>

</details>
Commit: 305e0cad28026ffb3633b4da90c501d39797cf1f

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| buildDuration | 12.3s | 12.8s | ⚠️ +520ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.196 | 2.224 | ⚠️ +0.03 |
| / avg req/sec | 1138.35 | 1124.23 | ⚠️ -14.12 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.459 | 1.486 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 1713.82 | 1682 | ⚠️ -31.82 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.79 kB | 5.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| index.html gzip | 955 B | 955 B | ✓ |
| link.html gzip | 962 B | 962 B | ✓ |
| withRouter.html gzip | 947 B | 947 B | ✓ |
| Overall change | 2.86 kB | 2.86 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| buildDuration | 13.6s | 13.5s | -61ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.79 kB | 5.79 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _error.js | 881 kB | 881 kB | ✓ |
| 404.html | 4.17 kB | 4.17 kB | ✓ |
| hooks.html | 3.79 kB | 3.79 kB | ✓ |
| index.js | 881 kB | 881 kB | ✓ |
| link.js | 923 kB | 923 kB | ✓ |
| routerDirect.js | 917 kB | 917 kB | ✓ |
| withRouter.js | 917 kB | 917 kB | ✓ |
| Overall change | 4.53 MB | 4.53 MB | ✓ |

</details>

</details>
Commit: 43dd874d707b9d0323863b7ae90dce4556ae8dda

**Review by developit — DISMISSED** [review]

**Comment by timneutkens** [comment]
Can you fix the linting?

**Comment by jackyef** [comment]
@timneutkens  Sorry, I was only making changes from github ui for this PR, so I didn't run the linters.

Pushed a new commit to run the linters, should be okay now.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| buildDuration | 12.2s | 12.6s | ⚠️ +382ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.036 | 2.031 | 0 |
| / avg req/sec | 1227.62 | 1231.07 | +3.45 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.233 | 1.225 | -0.01 |
| /error-in-render avg req/sec | 2027.65 | 2041.64 | +13.99 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.8 kB | 5.8 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| index.html gzip | 954 B | 954 B | ✓ |
| link.html gzip | 962 B | 962 B | ✓ |
| withRouter.html gzip | 948 B | 948 B | ✓ |
| Overall change | 2.86 kB | 2.86 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| buildDuration | 13.9s | 13.7s | -155ms |
| nodeModulesSize | 66.5 MB | 66.5 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.72 kB | 6.72 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..cead.js gzip | 9.77 kB | 9.77 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 56.4 kB | 56.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.8 kB | 5.8 kB | ✓ |
| webpack-HASH..dule.js gzip | 751 B | 751 B | ✓ |
| 19b7e98f51cc..dule.js gzip | 6.17 kB | 6.17 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 51.9 kB | 51.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.4 kB | 26.4 kB | ✓ |
| Overall change | 26.4 kB | 26.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 267 B | 267 B | ✓ |
| _buildManife..dule.js gzip | 273 B | 273 B | ✓ |
| Overall change | 540 B | 540 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | jackyef/next.js patch-1 | Change |
| - | - | - | - |
| _error.js | 882 kB | 882 kB | ✓ |
| 404.html | 4.17 kB | 4.17 kB | ✓ |
| hooks.html | 3.79 kB | 3.79 kB | ✓ |
| index.js | 882 kB | 882 kB | ✓ |
| link.js | 924 kB | 924 kB | ✓ |
| routerDirect.js | 918 kB | 918 kB | ✓ |
| withRouter.js | 918 kB | 918 kB | ✓ |
| Overall change | 4.53 MB | 4.53 MB | ✓ |

</details>

</details>
Commit: 1359df15f9244ae803df0b1b01b2156101c32b45

**Review by Timer — APPROVED** [review]

**Pull request merged**